# Técnicas de Limpeza e Tratamento de Valores Ausentes para Análise de Dados.

## Pacotes Python usados no projeto

In [1]:
!pip install -q -U watermark

In [2]:
# imports
import math
import sys, os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import warnings
warnings.filterwarnings('ignore')

In [5]:
%reload_ext watermark
%watermark -a "Leonardo da Silva Neves"

Author: Leonardo da Silva Neves



## Carregando os dados
https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

In [3]:
# Criando uma lista para identificar possíveis representações para valores ausentes
lista_labels_valores_ausentes = ["n/a", "na", "undefined"]

In [4]:
# Carregando o dataset com valores ausentes
# O parâmetro na_values identifica os argumentos de entrada como valores NaN. A ultilização
# deste parâmetro é valida quando se sabe de ante
dataset_dsa = pd.read_csv("dataset.csv", na_values= lista_labels_valores_ausentes)

In [8]:
# Shape
dataset_dsa.shape

(150001, 55)

In [10]:
# Amostra das primeiras colunas (Algumas colunas aparecem 'truncadas')
dataset_dsa.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0  9.16456699548519E+015  ...          15854611.0           2501332.0   
1                L77566A  ...          20247395.0          19111729.0   
2                D42335A  ...          19725661.0          14699576.0   
3                T21824A  ...          21388122.0          15146643.0   
4                D88865A  ...          15259380.0          18962873.0   

   Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0           8198936.0           9656251.0        278082303.0   
1          18338413.0          17227132.0        608750074.0   
2          17587794.0           6163408.0        229584621.0   
3          13994646.0           1097942.0        799538153.0   
4          17124581.0            415218.0        527707248.0   

   Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  \
0         14344150.0       171744450.0         8814393.0        36749741.0   
1          1170709.0       526904238.0        15055145.0        53800391.0   
2           395630.0       410692588.0         4215763.0        27883638.0   
3         10849722.0       749039933.0        12797283.0        43324218.0   
4          3529801.0       550709500.0        13910322.0        38542814.0   

   Total DL (Bytes)  
0       308879636.0  
1       653384965.0  
2       279807335.0  
3       846028530.0  
4       569138589.0  

[5 rows x 55 columns]

In [5]:
# Alteração de parâmetro do Pandas "display.max_columns" para almentar o número de colunas exibidas ao imprimir o dataframe.
pd.set_option('display.max_columns', 100)
# Caso o df tenha mais de 100 colunaas, o que passar será trucado.

In [13]:
dataset_dsa.tail()

Bearer Id           Start  Start ms              End  End ms  \
149996  7.277826e+18  4/29/2019 7:28     451.0   4/30/2019 6:02   214.0   
149997  7.349883e+18  4/29/2019 7:28     483.0  4/30/2019 10:41   187.0   
149998  1.311448e+19  4/29/2019 7:28     283.0  4/30/2019 10:46   810.0   
149999  1.311448e+19  4/29/2019 7:28     696.0  4/30/2019 10:40   327.0   
150000           NaN             NaN       NaN              NaN     NaN   

        Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
149996    81230.0  2.082022e+14   3.365069e+10  3.548311e+13   
149997    97970.0  2.082019e+14   3.366345e+10  3.566051e+13   
149998    98249.0  2.082017e+14   3.362189e+10  3.572121e+13   
149999    97910.0  2.082021e+14   3.361962e+10  8.618620e+13   
150000        NaN           NaN            NaN           NaN   

       Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
149996            D20434A             32.0              0.0   
149997            D10223C             27.0              2.0   
149998            T51102A             43.0              6.0   
149999            L88342B             37.0              5.0   
150000                NaN              NaN              NaN   

        Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
149996                     52.0                     65.0   
149997                     23.0                     54.0   
149998                     43.0                     47.0   
149999                     34.0                     37.0   
150000                      NaN                      NaN   

        TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
149996                          NaN                          NaN   
149997                          NaN                          NaN   
149998                          NaN                          NaN   
149999                          NaN                          NaN   
150000                          NaN                          NaN   

        DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
149996                100.0                             0.0   
149997                100.0                             0.0   
149998                100.0                             0.0   
149999                100.0                             0.0   
150000                  NaN                             NaN   

        250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  \
149996                            0.0                 0.0   
149997                            0.0                 0.0   
149998                            0.0                 0.0   
149999                            0.0                 0.0   
150000                            NaN                 NaN   

        UL TP < 10 Kbps (%)  10 Kbps < UL TP < 50 Kbps (%)  \
149996                100.0                            0.0   
149997                100.0                            0.0   
149998                100.0                            0.0   
149999                100.0                            0.0   
150000                  NaN                            NaN   

        50 Kbps < UL TP < 300 Kbps (%)  UL TP > 300 Kbps (%)  HTTP DL (Bytes)  \
149996                             0.0                   0.0              NaN   
149997                             0.0                   0.0              NaN   
149998                             0.0                   0.0              NaN   
149999                             0.0                   0.0              NaN   
150000                             NaN                   NaN              NaN   

        HTTP UL (Bytes)  Activity Duration DL (ms)  Activity Duration UL (ms)  \
149996              NaN                    42376.0                    41915.0   
149997              NaN                    17264.0                    16759.0   
149998              NaN                    25003.0                    28640.0   
149999              NaN                    13405.0                    34088.0   
150000         

In [6]:
# Carregando o dicionário de dados
dicionario = pd.read_excel('dicionario.xlsx')

In [15]:
# Shape
dicionario.shape

(56, 2)

In [17]:
# Amostra de dados
dicionario.head(15)
# Algumas descrições na coluna Description estão truncadas pelo tamanho do texto.

Fields                                        Description
0                 bearer id                             xDr session identifier
1                 Dur. (ms)                  Total Duration of the xDR (in ms)
2                     Start      Start time of the xDR (first frame timestamp)
3                  Start ms  Milliseconds offset of start time for the xDR ...
4                       End         End time of the xDR (last frame timestamp)
5                    End ms  Milliseconds offset of end time of the xDR (la...
6                  Dur. (s)                   Total Duration of the xDR (in s)
7                      IMSI           International Mobile Subscriber Identity
8             MSISDN/Number  MS International PSTN/ISDN Number of mobile - ...
9                      IMEI            International Mobile Equipment Identity
10       Last Location Name  User location call name (2G/3G/4G) at the end ...
11          Avg RTT DL (ms)  Average Round Trip Time measurement Downlink d...
12          Avg RTT UL (ms)  Average Round Trip Time measurement Uplink dir...
13  Avg Bearer TP DL (kbps)  Average Bearer Throughput for Downlink (kbps) ...
14  Avg Bearer TP UL (kbps)  Average Bearer Throughput for uplink (kbps) - ...

In [7]:
# Definindo um valor grande para a largura da coluna.
pd.set_option('display.max_colwidth', 100)

In [19]:
# Amostra de dados
dicionario.head(10)

Fields  \
0      bearer id   
1      Dur. (ms)   
2          Start   
3       Start ms   
4            End   
5         End ms   
6       Dur. (s)   
7           IMSI   
8  MSISDN/Number   
9           IMEI   

                                                             Description  
0                                                 xDr session identifier  
1                                      Total Duration of the xDR (in ms)  
2                          Start time of the xDR (first frame timestamp)  
3  Milliseconds offset of start time for the xDR (first frame timestamp)  
4                             End time of the xDR (last frame timestamp)  
5      Milliseconds offset of end time of the xDR (last frame timestamp)  
6                                       Total Duration of the xDR (in s)  
7                               International Mobile Subscriber Identity  
8          MS International PSTN/ISDN Number of mobile - customer number  
9                                International Mobile Equipment Identity

## Análise Exploratória

In [8]:
# info
dataset_dsa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

In [9]:
# Estatísticas descritivas
dataset_dsa.describe()

Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.490100e+05  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082016e+14   
std    2.893173e+18     288.611834     288.097653  8.103762e+04  2.148809e+10   
min    6.917538e+18       0.000000       0.000000  7.142000e+03  2.040471e+14   
25%    7.349883e+18     250.000000     251.000000  5.744050e+04  2.082014e+14   
50%    7.349883e+18     499.000000     500.000000  8.639900e+04  2.082015e+14   
75%    1.304243e+19     749.000000     750.000000  1.324302e+05  2.082018e+14   
max    1.318654e+19     999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847455e+13       109.795706        17.662883   
std     2.447443e+12  2.241637e+13       619.782739        84.793524   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365130e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611970e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
count            150000.000000            150000.000000   
mean              13300.045927              1770.428647   
std               23971.878541              4625.355500   
min                   0.000000                 0.000000   
25%                  43.000000                47.000000   
50%                  63.000000                63.000000   
75%               19710.750000              1120.000000   
max              378160.000000             58613.000000   

       TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
count                 6.185500e+04                 5.335200e+04   
mean                  2.080991e+07                 7.596587e+05   
std                   1.825665e+08                 2.645305e+07   
min                   2.000000e+00                 1.000000e+00   
25%                   3.565150e+04                 4.694750e+03   
50%                   5.687300e+05                 2.094950e+04   
75%                   3.768308e+06                 8.402025e+04   
max                   4.294426e+09                 2.908226e+09   

       DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
count        149247.000000                   149247.000000   
mean             92.844754                        3.069355   
std              13.038031                        6.215233   
min               0.000000                        0.000000   
25%              91.000000                        0.000000   
50%             100.000000                        0.000000   
75%             100.000000                        4.000000   
max             100.000000                       93.000000   

       250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
count                  149247.000000       149247.000000        149209.000000   
mean                        1.717341            1.609654            98.530142   
std                         4.159538            4.828890             4.634285   
min                         0.000000            0.000000             0.000000   
25%                         0.000000            0.000000            99.000000   
50%                         0.000000            0.000000           100.000000   
75%                         1.000000            0.000000           100.000000   
max                       100.000000           94.000000           100.000000   

       10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
count                  149209.000000                   149209.000000   
mean                        0.776749                        0.

Não faz sentido calcular estatísticas descritivas para variáveis Baread id, IMSI, MSISDN/Number e IMEI Embora essas colunas sejam
do tipo numérico, seus valores servem como identificadores. Mas o método describe() calcula as estatísticas de todas as colunas numéricas. Essas estatísticas estão sendo calculadas antes que os dados sejam limpos. Portanto, pode haver mudanças depois que os valores ausentes e outliers são tratados.

In [10]:
# Shape
dataset_dsa.shape

(150001, 55)

In [11]:
dicionario.shape

(56, 2)

Existem 150001 linhas e 55 colunas no dataframe. No entando, temos 56 colunas com seus nomes e descrições no dicionário.
Isso significa que há uma coluna descrita, mas não incluída no dataframe.

In [12]:
# Concatenar os datasets
df_compara_colunas = pd.concat([pd.Series(dataset_dsa.columns.tolist()), dicionario['Fields']], axis=1)

In [17]:
df_compara_colunas.columns

Index(['Coluna do Dataset', 'Coluna do Dicionário'], dtype='object')

In [16]:
# Renomeia as colunasc
df_compara_colunas.rename(columns = {0:'Coluna do Dataset', 'Fields': 'Coluna do Dicionário'}, inplace = True)

In [18]:
df_compara_colunas.head()

Coluna do Dataset Coluna do Dicionário
0         Bearer Id            bearer id
1             Start            Dur. (ms)
2          Start ms                Start
3               End             Start ms
4            End ms                  End